DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/hardware_reuse/naive/src/main.cpp

```diff
@@ -0,0 +1,152 @@
+#include <iostream>
+
+// oneAPI headers
+#include <sycl/ext/intel/fpga_extensions.hpp>
+#include <sycl/sycl.hpp>
+
+#include "exception_handler.hpp"
+
+// Forward declare the kernel name in the global scope. This is an FPGA best
+// practice that reduces name mangling in the optimization reports.
+class VectorOpID;
+
+constexpr int kVectSize = 3;
+constexpr int N = 5;
+using D3Vector = std::array<float, 3>;
+
+// Minimum capacity of a pipe.
+// Set to 0 to allow the compiler to save area if possible.
+constexpr size_t kPipeMinCapacity = 0;
+
+// Pipes
+class IDInputPipeA;
+class IDOutputPipeZ;
+
+using InputPipeA =
+    sycl::ext::intel::experimental::pipe<IDInputPipeA, D3Vector, kPipeMinCapacity>;
+using OutputPipeZ =
+    sycl::ext::intel::experimental::pipe<IDOutputPipeZ, float, 5>;
+
+// The square-root of a dot-product is an expensive operation.
+float OpSqrt(D3Vector val, const D3Vector coef)
+{
+  float res = sqrt(val[0] * coef[0] + val[1] * coef[1] + val[2] * coef[2]);
+  return res;
+}
+
+struct VectorOp
+{
+  int len;
+
+  void operator()() const
+  {
+    constexpr D3Vector coef1 = {0.2, 0.3, 0.4};
+    constexpr D3Vector coef2 = {0.6, 0.7, 0.8};
+
+    D3Vector new_item, item;
+    item = InputPipeA::read();
+    new_item[0] = OpSqrt(item, coef1);
+    item = InputPipeA::read();
+    new_item[1] = OpSqrt(item, coef1);
+    item = InputPipeA::read();
+    new_item[2] = OpSqrt(item, coef1);
+
+    OutputPipeZ::write(OpSqrt(new_item, coef2));
+  }
+};
+
+int main()
+{
+  bool passed = false;
+
+  try
+  {
+
+    // Use compile-time macros to select either:
+    //  - the FPGA emulator device (CPU emulation of the FPGA)
+    //  - the FPGA device (a real FPGA)
+    //  - the simulator device
+#if FPGA_SIMULATOR
+    auto selector = sycl::ext::intel::fpga_simulator_selector_v;
+#elif FPGA_HARDWARE
+    auto selector = sycl::ext::intel::fpga_selector_v;
+#else // #if FPGA_EMULATOR
+    auto selector = sycl::ext::intel::fpga_emulator_selector_v;
+#endif
+
+    sycl::queue q(selector, fpga_tools::exception_handler,
+                  sycl::property::queue::enable_profiling{});
+
+    auto device = q.get_device();
+
+    std::cout << "Running on device: "
+              << device.get_info<sycl::info::device::name>().c_str()
+              << std::endl;
+
+    // initialize input D3Vector
+    constexpr float test_vecs[kVectSize][D3Vector{}.size()] = {
```

Comment:
```suggestion
    constexpr float test_vecs[D3Vector{}.size()][D3Vector{}.size()] = {
```
having two variables that mean the same thing is confusing 😊